library/core/src/ptr/mod.rs

```diff
@@ -1451,21 +1451,62 @@ macro_rules! fnptr_impls_safety_abi {
     }
 }
 
-macro_rules! fnptr_impls_args {
-    ($($Arg: ident),+) => {
-        fnptr_impls_safety_abi! { extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "Rust" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "C" fn($($Arg),+) -> Ret, $($Arg),+ }
-        fnptr_impls_safety_abi! { unsafe extern "C" fn($($Arg),+ , ...) -> Ret, $($Arg),+ }
+macro_rules! fnptr_impl_for_abi {
+    ($Abi:tt, $($Arg: ident),*) => {
+        fnptr_impls_safety_abi! { extern $Abi fn($($Arg),*) -> Ret, $($Arg),* }
+        fnptr_impls_safety_abi! { unsafe extern $Abi fn($($Arg),*) -> Ret, $($Arg),* }
     };
-    () => {
+}
+
+macro_rules! fnptr_impl_for_abi_variadic_support {
+    ($Abi:tt, ) => {
         // No variadic functions with 0 parameters
-        fnptr_impls_safety_abi! { extern "Rust" fn() -> Ret, }
-        fnptr_impls_safety_abi! { extern "C" fn() -> Ret, }
-        fnptr_impls_safety_abi! { unsafe extern "Rust" fn() -> Ret, }
-        fnptr_impls_safety_abi! { unsafe extern "C" fn() -> Ret, }
+        fnptr_impls_safety_abi! { extern $Abi fn() -> Ret, }
+        fnptr_impls_safety_abi! { unsafe extern $Abi fn() -> Ret, }
+    };
+
+    ($Abi:tt, $($Arg: ident),*) => {
+        fnptr_impl_for_abi! { $Abi, $($Arg),* }
+
+        fnptr_impls_safety_abi! { extern $Abi fn($($Arg),*, ...) -> Ret, $($Arg),* }
+        fnptr_impls_safety_abi! { unsafe extern $Abi fn($($Arg),*, ...) -> Ret, $($Arg),* }
+    }
+}
+
+macro_rules! fnptr_impls_args {
+    ($($Arg: ident),*) => {
+        fnptr_impl_for_abi! { "Rust", $($Arg),* }
+        fnptr_impl_for_abi! { "stdcall", $($Arg),* }
+        fnptr_impl_for_abi! { "fastcall", $($Arg),* }
+        fnptr_impl_for_abi! { "aapcs", $($Arg),* }
+        fnptr_impl_for_abi! { "win64", $($Arg),* }
+        fnptr_impl_for_abi! { "sysv64", $($Arg),* }
+        fnptr_impl_for_abi! { "system", $($Arg),* }
+        fnptr_impl_for_abi! { "rust-intrinsic", $($Arg),* }
+        fnptr_impl_for_abi! { "rust-call", $($Arg),* }
+        fnptr_impl_for_abi! { "platform-intrinsic", $($Arg),* }
+        fnptr_impl_for_abi! { "unadjusted", $($Arg),* }
+
+        // fnptr_impl_for_abi! { "C-unwind", $($Arg),* }
```

Comment:
Drafting until I either know how to get this to work, or know that it's not possible.

When these are enabled, the compiler complains about the stability. If I add `#[unstable()]` etc etc to the traits, I get ineffective_unstable_trait_impl.

Reply:
cc @rust-lang/compiler for a proper answer to this but if I had to guess my suspicion is that the mechanism used for gating the unstable calling conventions is separate from the normal one for stability of items in the language. Trait impls can't be unstable if the type they're associated with is stable, which is where the lint you're triggering is from. When the compiler goes to check if the trait definition is being implemented for an unstable item it might be assuming its stable because it only sees the function pointer rather than the calling convention. If that's the case then this is a compiler bug but it may take some non-trivial refactoring to fix.

Reply:
In the case that it's not possible, would it be preferred for me to remove the commented code, or leave it with a note as to why it's commented out?

Reply:
Why not enable the `#[feature]`s in question and implement the trait unconditionally? 

I believe one way or another trying to use the unstable calling convention will require a feature in the user code, regardless of whether these traits are implemented or not.

Reply:
I'm unsure on what you're asking me to do here, is rustc allowed to enable `#[feature]`?

Reply:
I'm not asking anything in particular, just throwing the idea out into the pool.

And yes, the standard library implementation [is allowed to enable `#[feature]`s]( https://github.com/rust-lang/rust/blob/f624427f8771c00819684c783bb841bf72095704/library/core/src/lib.rs#L97-L216).

Reply:
If that's seen as normal then I can do that.